src/lib.rs

```diff
@@ -0,0 +1,54 @@
+use serde_json_core;
+
+pub trait StringSet {
+    fn string_set(&mut self, topic_parts:
+            core::iter::Peekable<core::str::Split<char>>, value: &str) ->
+            Result<(),()>;
+
+}
+
+macro_rules! derive_primitive {
+    ($x:ty) => {
+        derive_single!($x);
+        // This is needed until const generics is stabilized https://github.com/rust-lang/rust/issues/44580
+        derive_array!($x, (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32));
+    }
+}
+macro_rules! derive_single {
+    ($x:ty) => {
+        impl StringSet for $x {
+            fn string_set(&mut self, _topic_parts:
+                core::iter::Peekable<core::str::Split<char>>, value: &str) ->
+                Result<(),()> {
+                *self = serde_json_core::from_str(value)
```

Comment:
This should fail if `_topic_parts` contains more parts. This should be a leaf node and there should be no remaining topic string at this point. As written, this will still succeed even if we requested a sub-field of a primitive type, but it should error.

Reply:
@irwineffect We should fix this before merging, as this could result in incorrect behavior still

Reply:
I now realize this will disappear with field annotations